### PR TITLE
Add averaged perceptron cookbook page

### PR DESCRIPTION
**Pull request by geektoni**

### CODE REVIEW

File: doc/cookbook/source/examples/binary_classifier/averaged_perceptron.rst [hunk 1]
```diff
@@ -0,0 +1,61 @@
+===================
+Averaged Perceptron
+===================
+
+The Averaged Perceptron is an online binary classifier which is used to determine whether an input,
```
Comment (karlnapf) [inline doc/cookbook/source/examples/binary_classifier/averaged_perceptron.rst]: don't capitalise words that are non-names, like "averaged"

Comment (karlnapf) [inline doc/cookbook/source/examples/binary_classifier/averaged_perceptron.rst]: Don't explain what classification is ;) This kind of stuff is more for the notebooks / longer examples. This is supposed to be a minimal.
The averaged Perceptron is an online binary classifier.

File: doc/cookbook/source/examples/binary_classifier/averaged_perceptron.rst [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+namely a feature vector :math:`\mathbf{x}`, belongs to some specific class or not. It is an extension
+of the simple perceptron algorithm because it uses the `average weight` vector
```
Comment (karlnapf) [inline doc/cookbook/source/examples/binary_classifier/averaged_perceptron.rst]: Make it simpler: "It is an extension of the standard Perceptron; it uses the averaged weight and bias."

File: doc/cookbook/source/examples/binary_classifier/averaged_perceptron.rst [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+and the `average bias` to classify the given inputs.
+The predicted class :math:`c` is given by the function below:
```
Comment (karlnapf) [inline doc/cookbook/source/examples/binary_classifier/averaged_perceptron.rst]: Given a vector :math:`\mathbf{x}`, the predicted class :math:`c` is given by

File: doc/cookbook/source/examples/binary_classifier/averaged_perceptron.rst [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+
+.. math::
+
+    c = \theta\left(\mathbf{w} \cdot \mathbf{x}+b\right)
```
Comment (karlnapf) [inline doc/cookbook/source/examples/binary_classifier/averaged_perceptron.rst]: Do you actually use the c anywhere? If not, remove the name altogether

File: doc/cookbook/source/examples/binary_classifier/averaged_perceptron.rst [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+
+Where :math:`\mathbf{w}` is the average weight vector, :math:`\mathbf{x}` is the input vector we want to classify,
```
Comment (karlnapf) [inline doc/cookbook/source/examples/binary_classifier/averaged_perceptron.rst]: Grammar error, this is either the same sentence, or a new one (no capital W)

Better:
"Here, w is the average weight vector and b is the average bias. and \theta is a step function:

File: doc/cookbook/source/examples/binary_classifier/averaged_perceptron.rst [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+.. math::
+
+    \theta(x) =
+    \begin{cases}
+    1 & x > 0 \\
+    0 & x = 0 \\
+    -1 & x < 0
+    \end{cases}
+
+See Chapter 17 in :cite:`barber2012bayesian` for a brief explanation of the perceptron.
```
Comment (karlnapf) [inline doc/cookbook/source/examples/binary_classifier/averaged_perceptron.rst]: Perceptron

File: doc/cookbook/source/examples/binary_classifier/averaged_perceptron.rst [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+-------
+Example
+-------
+
+Given a linearly separable dataset, we create some CDenseFeatures
+(RealFeatures, here 64 bit float values) and some :sgclass:`CBinaryLabels` to set up the training and validation sets.
+
+.. sgexample:: averaged_perceptron.sg:create_features
+
+Secondly, we create the :sgclass:`CAveragedPerceptron` instance by passing it the traning features and labels.
```
Comment (karlnapf) [inline doc/cookbook/source/examples/binary_classifier/averaged_perceptron.rst]: Remove "secondly"

File: doc/cookbook/source/examples/binary_classifier/averaged_perceptron.rst [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+We also set its learn rate and its maximum iterations.
+
+.. sgexample:: averaged_perceptron.sg:set_parameters
+
+Then we train the :sgclass:`CAveragedPerceptron` and we apply it to the test data, which gives the predicted :sgclass:`CLabels`.
```
Comment (karlnapf) [inline doc/cookbook/source/examples/binary_classifier/averaged_perceptron.rst]: what kind of labels does it return? I guess BinaryLabels, should be that sgclass

File: doc/cookbook/source/examples/binary_classifier/averaged_perceptron.rst [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+
+.. sgexample:: averaged_perceptron.sg:train_and_apply
+
+We can now also extract :math:`\mathbf{w}` and :math:`b`.
```
Comment (karlnapf) [inline doc/cookbook/source/examples/binary_classifier/averaged_perceptron.rst]: We can also extract the average weight w and the bias b.

(reference the letters introduced above), shorten (no "now")

File: doc/cookbook/source/examples/binary_classifier/averaged_perceptron.rst [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+
+.. sgexample:: averaged_perceptron.sg:extract_weights
+
+Finally, we can evaluate the performance of our model, e.g. using :sgclass:`CAccuracyMeasure`.
```
Comment (karlnapf) [inline doc/cookbook/source/examples/binary_classifier/averaged_perceptron.rst]: Finally, we can evaluate the performance, e.g. using :sgclass:`CAccuracyMeasure`.

File: doc/cookbook/source/examples/binary_classifier/averaged_perceptron.rst (hunk 6 above)
Comment (karlnapf) [inline doc/cookbook/source/examples/binary_classifier/averaged_perceptron.rst]: chapter, not Chapter

**Comment by geektoni** [comment]
Thanks for your in-depth analysis @karlnapf! I corrected the cookbook page as you suggested.

File: doc/cookbook/source/examples/binary_classifier/averaged_perceptron.rst [hunk 11]
```diff
@@ -0,0 +1,60 @@
+===================
+Averaged Perceptron
+===================
+
+The averaged Perceptron is an online binary classifier. It is an extension
+of the standard Perceptron algorithm; it uses the `averaged weight` and
+`bias`.
+Given a vector :math:`\mathbf{x}`, the predicted class :math:`c` is given by:
```
Comment (karlnapf) [inline doc/cookbook/source/examples/binary_classifier/averaged_perceptron.rst]: No need for c here if you dont use it downstream

File: doc/cookbook/source/examples/binary_classifier/averaged_perceptron.rst [hunk 12]
```diff
@@ -0,0 +1,60 @@
+===================
+Averaged Perceptron
+===================
+
+The averaged Perceptron is an online binary classifier. It is an extension
+of the standard Perceptron algorithm; it uses the `averaged weight` and
```
Comment (karlnapf) [inline doc/cookbook/source/examples/binary_classifier/averaged_perceptron.rst]: `averaged` weight

**Comment by geektoni** [comment]
Updated @karlnapf

**Comment by karlnapf** [comment]
Thanks!

**Pull request merged**